executor/src/witgen/query_processor.rs

```diff
@@ -103,102 +96,33 @@ struct Symbols<'a, T: FieldElement> {
-                        }
-                        _ => panic!(
-                            "Arrays and queries should have been found by try_column_by_name()"
-                        ),
-                    }
-                }
-                None => Err(EvalError::SymbolNotFound(format!(
-                    "Symbol {name} not found."
-                ))),
-            },
-        }
+    ) -> Result<Value<'a, T, NoCustom>, EvalError> {
+        Definitions(&self.fixed_data.analyzed.definitions).lookup(name, generic_args)
     }
-    fn eval_function_application(
-        &self,
-        function: Reference<'a>,
-        arguments: &[Rc<Value<'a, T, Reference<'a>>>],
-    ) -> Result<Value<'a, T, Reference<'a>>, EvalError> {
-        if arguments.len() != 1 {
-            Err(EvalError::TypeError(format!(
-                "Expected one argument, but got {}",
-                arguments.len()
-            )))?
-        };
-        let Value::Integer(row) = arguments[0].as_ref() else {
+
+    fn eval_expr(&self, expr: AlgebraicExpression<T>) -> Result<Value<'a, T, NoCustom>, EvalError> {
+        let AlgebraicExpression::Reference(poly_ref) = expr else {
             return Err(EvalError::TypeError(format!(
-                "Expected integer but got {}",
-                arguments[0]
+                "Can use std::prover::eval only directly on columns - tried to evaluate {expr}"
             )));
         };
-        // TODO change this mechanism to use a built-in "eval" function instead.
-        Ok(Value::FieldElement(match function.poly_id.ptype {
-            PolynomialType::Committed | PolynomialType::Intermediate => {
-                let row_index = RowIndex::from_degree(
-                    DegreeType::try_from(row).unwrap(),
-                    self.fixed_data.degree,
-                );
-                let next = self.rows.is_row_number_next(row_index).map_err(|_| {
-                    EvalError::OutOfBounds(format!("Referenced row outside of window: {row}"))
-                })?;
-                let poly_ref = AlgebraicReference {
-                    name: function.name.to_string(),
-                    poly_id: function.poly_id,
-                    next,
-                };
 
-                self.rows
-                    .get_value(&poly_ref)
-                    .ok_or(EvalError::DataNotAvailable)?
-            }
+        Ok(Value::FieldElement(match poly_ref.poly_id.ptype {
+            PolynomialType::Committed | PolynomialType::Intermediate => self
+                .rows
+                .get_value(&poly_ref)
+                .ok_or(EvalError::DataNotAvailable)?,
             PolynomialType::Constant => {
-                let values = self.fixed_data.fixed_cols[&function.poly_id].values;
-                values[(usize::try_from(row).unwrap() % values.len())
-                    .to_u64()
-                    .unwrap() as usize]
+                // TODO do we need this special case?
```

Comment:
@georgwiese do you know if we can somehow access the fixed column values directyl from a row pair? Or do I need to go through fixed_data like here?

Reply:
As far as I can see, `RowPair::get_value` does not currently handle reference to fixed columns, but I think it should, as it already has access to `fixed_data`! `RowPair::evaluate` already handles this (but it's evaluating to an `AffineExpression`, which I guess we don't want to do here).

Reply:
I mean it depends. I'll remove the TODO here and keep it like that for now.